packages/shared/src/governance.ts

```diff
@@ -0,0 +1,897 @@
+import { BN } from 'bn.js'
+import { assert, describe, test } from 'vitest'
+
+import { Chain, defaultAccounts } from '@e2e-test/networks'
+import { setupNetworks } from '@e2e-test/shared'
+import { check, checkEvents } from './helpers/index.js'
+
+import { sendTransaction } from '@acala-network/chopsticks-testing'
+
+import {
+  FrameSupportPreimagesBounded,
+  FrameSupportScheduleDispatchTime,
+  KitchensinkRuntimeOriginCaller,
+  PalletConvictionVotingTally,
+  PalletConvictionVotingVoteCasting,
+  PalletConvictionVotingVoteVoting,
+  PalletReferendaDecidingStatus,
+  PalletReferendaDeposit,
+  PalletReferendaReferendumInfoConvictionVotingTally,
+  PalletReferendaReferendumStatusConvictionVotingTally,
+} from '@polkadot/types/lookup'
+import { ITuple } from '@polkadot/types/types'
+import { Option, bool, u16, u32 } from '@polkadot/types'
+import { encodeAddress } from '@polkadot/util-crypto'
+
+/**
+ * Ongoing referenda are stored as `PalletReferendaReferendumStatusConvictionVotingTally`, an
+ * interface in PJS.
+ *
+ * In TypeScript, it is not possible to get a list of an interface's properties.
+ *
+ * In order to get properties to then granularly compare the same referenda in different
+ * stages, the below class is required, to then instantiate as a PJS's interface.
+ */
+class OngoingReferendumStatus {
+  readonly track!: u16
+  readonly origin!: KitchensinkRuntimeOriginCaller
+  readonly proposal!: FrameSupportPreimagesBounded
+  readonly enactment!: FrameSupportScheduleDispatchTime
+  readonly submitted!: u32
+  readonly submissionDeposit!: PalletReferendaDeposit
+  readonly decisionDeposit!: Option<PalletReferendaDeposit>
+  readonly deciding!: Option<PalletReferendaDecidingStatus>
+  readonly tally!: PalletConvictionVotingTally
+  readonly inQueue!: bool
+  readonly alarm!: Option<ITuple<[u32, ITuple<[u32, u32]>]>>
+}
+
+/**
+ * Compare the selected properties of two referenda.
+ *
+ * Fails if any of the properties to be compared is different.
+ *
+ * When awaiting a referendum's preparation period, it is desirable to compare the referendum
+ * pre- and post-block-execution; in this case, an optional error message parameter is passable,
+ * to allow indicating the offending iteration.
+ *
+ * @param ref1
+ * @param ref2
+ * @param propertiesToBeSkipped List of properties to be skipped in the comparison
+ * @param errorMsg Additional error message to use when using this function inside a loop, to
+ *        identify failing iteration.
+ */
+function referendumCmp(
+  ref1: PalletReferendaReferendumStatusConvictionVotingTally,
+  ref2: PalletReferendaReferendumStatusConvictionVotingTally,
+  propertiesToBeSkipped: string[],
+  errorMsg?: string,
+) {
+  type ReferendumProperties = (keyof PalletReferendaReferendumStatusConvictionVotingTally)[]
```

Comment:
In these tests, it is desirable (in my opinion) to verify that extrinsics which alter the state of a referendum change only portions relevant to them, leaving the object equal otherwise:
* a decision deposit only changes that field (and the `alarm`), and nothing else.
* voting only changes the tally (and the `alarm`)

In order to compare a variable (and dynamic) list of properties on an object, this is the reusable solution I could come up with.

Reply:
ok but you still don't need to create a class just to get its keys?
```ts
const properties: ReferendumProperties[] = [
"track", "origin", // ...
]
```

will do the same thing? and with type checking (invalid property key should result an error), and less duplicated code (the property type)

and still not sure why you can't do
```ts
const properties = Object.keys(ref1) // or ref1.toJSON()
```

Reply:
In order to compare each of the pertinent properties of `PalletReferendaReferendumStatusConvictionVotingTally` of the two referenda, [Polkadot.js'](https://polkadot.js.org/docs/api/start/types.basics/#everything-is-a-type) `.eq` is used.

If the comparison function were to become

```typescript
function referendumCmp(
  ref1: PalletReferendaReferendumStatusConvictionVotingTally,
  ref2: PalletReferendaReferendumStatusConvictionVotingTally,
  propertiesToBeSkipped: string[],
  errorMsg?: string,
) {
  const properties: string[] = Object.keys(ref1)

  properties
    .filter((prop) => !propertiesToBeSkipped.includes(prop as string))
    .forEach((prop) => {
      const cmp = ref1[prop].eq(ref2[prop]) <--- this won't work: "TypeError: ref1[prop].eq is not a function"

    ...
    }
...
}
```

(I tried it, the compiler error `"TypeError: ref1[prop].eq is not a function"` is the result)

Using the mock interface and class allows TS type inference to know that `ref1[prop].eq` will only ever be invoked for strings which all have `.eq` available:
1. `type ReferendumProperties = (keyof PalletReferendaReferendumStatusConvictionVotingTally)[]`
2. `const properties: ReferendumProperties = Object.keys(new OngoingReferendumStatus()) as ReferendumProperties`

Reply:
how about `const properties: keyof PalletReferendaReferendumStatusConvictionVotingTally[] = Object.keys(ref1)`
you can also do `if ('eq' in ref1[prop])` to ensure every field have `eq`

another way is just `.toJSON` both ref1 and ref2 and `expect(ref1[prop]).toEqual(ref2[prop])`

Reply:
`const properties: keyof PalletReferendaReferendumStatusConvictionVotingTally[] = Object.keys(ref1)` fails with the compile time error `Type 'string[]' is not assignable to type 'keyof PalletReferendaReferendumStatusConvictionVotingTally[]'.`

`if ('eq' in ref1[prop])` fails with `Cannot use 'in' operator to search for 'eq' in undefined`.

In the end, I fell back to `const properties = ['track', 'origin', 'proposal', ...]`, which worked.
Out of curiosity, I also tried the `toJSON` method, and that also worked - I'll stick to `.eq`, if that's ok with you.